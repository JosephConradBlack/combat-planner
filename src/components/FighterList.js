import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { Segment } from 'semantic-ui-react';
import posed, { PoseGroup } from 'react-pose';

import _ from 'underscore';

import Fighter from './Fighter';

const MAX_INT = -9007199254740991;

const StyledSegmentGroup = styled(Segment.Group)`
  &&& {
    margin-bottom: 30px;
  }
`;

// Override Segment styling to re-implement Segment.Group>Segment styling after it is interrupted
//  by the intervening posed Item
const StyledFighter = styled(Fighter)`
  &&&& {
    border-top: none;
    margin-top: 0;
    bottom: 0;
    margin-bottom: 0;
    top: 0;
    border: none;
    
    border-radius: ${props =>
      props.index !== 0 ? '0' : '.28571429rem .28571429rem 0 0'}
    border-top:  ${props =>
      props.index === 0 ? '0' : '1px solid rgba(34, 36, 38, 0.15)'};
    border-radius: ${props =>
      props.index !== props.lastIndex ? '0' : '0 0 0.28571429rem 0.28571429rem'}
  }
`;

const prefilItemsWithPlaceholders = items => {
  if (items.length > 3) return items;

  return _
    .range(0, 3 - items.length)
    .map(num => ({ id: num, placeholder: true }))
    .concat(items);
};

const orderItems = items => {
  return _.sortBy(
    items,
    item => -1 * (item.initiative ? item.initiative : MAX_INT)
  );
};

const getItems = (items, lastIndex, onFighterUpdate) =>
  items.map((item, index) => (
    <Item key={item.id}>
      {item.placeholder ? (
        <StyledFighter key={item.id} />
      ) : (
        <StyledFighter
          key={item.id}
          index={index}
          lastIndex={lastIndex}
          item={item}
          handleChange={onFighterUpdate}
        />
      )}
    </Item>
  ));

const Item = posed.div({
  enter: { opacity: 1 },
  exit: { opacity: 0 }
});

const FighterList = ({ items, onFighterUpdate }) => {
  const itemsWithPlaceholders = prefilItemsWithPlaceholders(orderItems(items));
  const lastIndex = itemsWithPlaceholders.length - 1;

  return (
    <StyledSegmentGroup>
      <PoseGroup>
        {getItems(itemsWithPlaceholders, lastIndex, onFighterUpdate)}
      </PoseGroup>
    </StyledSegmentGroup>
  );
};

FighterList.propTypes = {
  items: PropTypes.array,
  onFighterUpdate: PropTypes.func
};

FighterList.defaultProps = {
  items: []
};

export default FighterList;
