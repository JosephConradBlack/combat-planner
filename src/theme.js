export default {
  h1: {
    fontSize: '6em',
    fontWeight: 300,
    letterSpacing: '0.06em',
    subheader: {
      fontSize: '1em',
      fontWeight: 'lighter'
    }
  }
};
