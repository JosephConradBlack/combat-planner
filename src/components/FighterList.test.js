import React from 'react';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

import _ from 'underscore';

import theme from '../theme';
import FighterList from './FighterList';

describe('FighterList', () => {
  it('renders with no props', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <FighterList />
      </ThemeProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with empty items array', () => {
    const items = [];
    const component = create(
      <ThemeProvider theme={theme}>
        <FighterList items={items} />
      </ThemeProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  const createComponentWithNFighters = numFighters => {
    let items = [];

    _.times(numFighters, idx => {
      items.push({
        id: idx,
        name: 'Item: ' + idx
      });
    });

    return create(
      <ThemeProvider theme={theme}>
        <FighterList items={items} />
      </ThemeProvider>
    );
  };

  it('renders with one item', () => {
    const component = createComponentWithNFighters(1);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with two items', () => {
    const component = createComponentWithNFighters(2);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with three items', () => {
    const component = createComponentWithNFighters(3);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with four items', () => {
    const component = createComponentWithNFighters(4);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
