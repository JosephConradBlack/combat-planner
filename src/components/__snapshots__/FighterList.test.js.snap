// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FighterList renders with empty items array 1`] = `
<div
  className="ui basic segment sc-bZQynM gXFCxG"
>
  <div
    className="ui container sc-gzVnrw dlvNYg"
  >
    <div
      className="ui segments"
    />
  </div>
</div>
`;

exports[`FighterList renders with no props 1`] = `
<div
  className="ui basic segment sc-bZQynM gXFCxG"
>
  <div
    className="ui container sc-gzVnrw dlvNYg"
  >
    <div
      className="ui segments"
    />
  </div>
</div>
`;

exports[`FighterList renders with two items 1`] = `
<div
  className="ui basic segment sc-bZQynM gXFCxG"
>
  <div
    className="ui container sc-gzVnrw dlvNYg"
  >
    <div
      className="ui segments"
    >
      <div
        className="ui segment"
      >
        <div
          className="item"
        >
          <div
            className="content"
          >
            <div
              className="extra"
            >
              <div
                className="ui left floated header"
              >
                1
              </div>
            </div>
            <div
              className="header sc-bxivhb bYOrRR"
            >
              Name
            </div>
            <div
              className="extra sc-ifAKCX kAHmWX"
            >
              <div>
                <div
                  className="extra sc-bdVaJa cdKNHG"
                >
                  <i
                    aria-hidden="true"
                    className="red heart outline big icon sc-bwzfXH cXPuzX"
                  />
                  <div
                    className="sc-htpNat iPQVHr"
                    color="red"
                  >
                    0
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
